jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Utf8Appendable.java

```diff
@@ -89,19 +89,47 @@ public Utf8Appendable(Appendable appendable)
 
     public abstract int length();
 
-    protected void reset()
+    /**
+     * Reset the appendable pass in {@link Utf8Appendable#Utf8Appendable(Appendable)}
+     */
+    protected abstract void resetAppendable();
+
+    /**
+     * @return {@code True} if the characters decoded have contained UTF8 coding errors.
+     */
+    public boolean hasCodingErrors()
+    {
+        return _codingErrors;
+    }
+
+    /**
+     * Reset the appendable, clearing the buffer, resetting decoding state and clearing any errors.
+     */
+    public void reset()
     {
         _state = UTF8_ACCEPT;
+        _codingErrors = false;
+        resetAppendable();
+    }
+
+    /**
+     * Partially reset the appendable: clear the buffer and clear any errors, but retain the decoding state
+     * of any partially decoded sequences.
+     */
+    public void partialReset()
+    {
+        int save = _state;
+        reset();
+        _state = save;
     }
 
     private void checkCharAppend() throws IOException
     {
         if (_state != UTF8_ACCEPT)
         {
             _appendable.append(REPLACEMENT);
-            int state = _state;
             _state = UTF8_ACCEPT;
-            throw new NotUtf8Exception("char appended in state " + state);
+            _codingErrors = true;
```

Comment:
This method needs work.
This will now add a replacement character, then the caller will add it's own (bad) character.

Reply:
Isn't the replacement character for the incomplete sequence, then the caller adds it's own character, which can't be bad as it is already a character.

Reply:
This means if you repeatedly pass the output back into Utf8Appendable the string grows, which is bad.
Imagine back/forth requests/responses like this.

Reply:
But we want it to grow. Somebody had called a method like append(char). We want to spend that char, but if we are incomplete then we append a replacement first.

Reply:
I think you misunderstand what I mean.

Take this example.

``` java
Utf8StringBuilder builder = new Utf8StringBuilder();
builder.append(stringFromRequest);
builder.complete();
String stringUsedInResponse = builder.toString();
```

Imagine that the client takes `stringUsedInResponse` and sends it back as `stringFromRequest` (don't get hung up on where in the request/response).
If that round trip makes the string grow every time then we have a bug.
The replacement character behavior should result in no growth of the string length when it is passed through the algorithm again.

Reply:
Imagine a string, supplied by an HTTP client, initially passed into Jetty as bad, then "sanitized" (with replacement characters).

We give that string back to the client (for whatever reason), in the "sanitized" form.

That same client gives us back that "sanitized" string in a different request.

That's the loop, a new Utf8StringBuilder each time, but the string on output is used again on a new Utf8StringBuilder (in append + toString)
That loop shouldn't result in a different String, or a longer String, as the replacement character should have sanitized it.
If each iteration through Utf8StringBuilder results in a different string, we have a bug.